### PR TITLE
pkg/ipmi: Restructure, rework and test coverage

### PR DESCRIPTION
**Pull request by ChriMarMe**

### CODE REVIEW

File: pkg/ipmi/structures.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 the u-root Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/ipmi/structures.go:1]:
```suggestion
// Copyright 2021 the u-root Authors. All rights reserved
```

File: pkg/ipmi/ipmi_dev.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 the u-root Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:1]: 2021

**Comment by rminnich** [comment]
quick question: this is 12 commits. does it make sense to squash into fewer commits? Mainly asking, I really don't know.

File: pkg/ipmi/ipmi_dev.go (hunk 2 above)
Comment (rminnich) [inline pkg/ipmi/ipmi_dev.go:1]: The general guidance on names like this is
pkg/ipmi/dev.go
as the name pkg/ipmi/ipmi_dev.go "stutters", i.e. repeats a name it need not repeat. I'd suggest changing this to dev.

Comment (rminnich) [inline pkg/ipmi/ipmi_dev.go:1]: double check copyright dates, 2019-2021 etc.

File: pkg/ipmi/ipmi_dev_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 the u-root Authors. All rights reserved
```
Comment (rminnich) [inline pkg/ipmi/ipmi_dev_test.go:1]: pkg/ipmi/dev_test.go

File: pkg/ipmi/ipmi_handler.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2019 the u-root Authors. All rights reserved
```
Comment (rminnich) [inline pkg/ipmi/ipmi_handler.go:1]: handler.go

File: pkg/ipmi/ipmi_linux.go [hunk 5]
```diff
@@ -18,5 +18,5 @@ func Open(devnum int) (*IPMI, error) {
 		return nil, err
```
Comment (rminnich) [inline pkg/ipmi/ipmi_linux.go:18]: I would try ipmi/linux.go and if that does not work to restrict its use, add a build tag.

File: pkg/ipmi/ipmi_mock.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2019 the u-root Authors. All rights reserved
```
Comment (rminnich) [inline pkg/ipmi/ipmi_mock.go:1]: mock.go

**Review by rminnich — COMMENTED** [review]
This is really nice, I only had a few nits, hope they are helpful.

File: pkg/ipmi/ipmi_dev.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	fileSetReadDeadline = func(f *os.File, t time.Duration) error {
+		return f.SetReadDeadline(time.Now().Add(t))
+	}
+	// connRead gives the option to overwrite the actual function call with a dummy function when writing tests.
+	connRead = func(f func(fd uintptr) bool, conn syscall.RawConn) error {
+		return conn.Read(f)
+	}
+)
+
+func (d *Dev) SendRequest(req *request) error {
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:42]: Add function comment.

File: pkg/ipmi/ipmi_dev.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	_, _, err := unixSyscall(unix.SYS_IOCTL, d.File.Fd(), _IPMICTL_SEND_COMMAND, uintptr(unsafe.Pointer(req)))
+	runtime.KeepAlive(req)
+	if err != 0 {
+		return fmt.Errorf("syscall failed with: %v", err.Error())
+	}
+	return nil
+}
+
+func (d *Dev) ReceiveResponse(msgID int64, resp *response, buf []byte) ([]byte, error) {
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:51]: Add function comment.

File: pkg/ipmi/ipmi_dev.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		return nil, fmt.Errorf("failed to set read deadline: %v", err)
+	}
+	if err := connRead(readMsg, conn); err != nil {
+		return nil, fmt.Errorf("failed to read rawconn: %v", err)
+	}
+
+	return result, rerr
+}
+
+func (d *Dev) GetFile() *os.File {
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:93]: Add function comment.

File: pkg/ipmi/ipmi_dev.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	// fileSyscallConn gives the option to overwrite the actual function call with a dummy function when writing tests.
+	fileSyscallConn = func(f *os.File) (syscall.RawConn, error) {
+		return f.SyscallConn()
+	}
+
+	// fileSetReadDeadline gives the option to overwrite the actual function call with a dummy function when writing tests.
+	fileSetReadDeadline = func(f *os.File, t time.Duration) error {
+		return f.SetReadDeadline(time.Now().Add(t))
+	}
+	// connRead gives the option to overwrite the actual function call with a dummy function when writing tests.
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:36]: Add newline before this line for consistency.

File: pkg/ipmi/ipmi_handler.go (hunk 4 above)
Comment (rjoleary) [inline pkg/ipmi/ipmi_handler.go:1]: 2021

File: pkg/ipmi/ipmi_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+func TestWatchdogRunning(t *testing.T) {
+	i := GetMockIPMI()
+	_, err := i.WatchdogRunning()
+	if err != nil {
+		t.Error(err)
+	}
+
+}
+
+func TestShutiffWatchdog(t *testing.T) {
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_test.go:20]:
```suggestion
func TestShutoffWatchdog(t *testing.T) {
```

File: pkg/ipmi/structures.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+type OEMTsEvent struct {
+	Timestamp        uint32
+	ManfID           [3]uint8
+	OEMTsDefinedData [6]uint8
+}
+
+// OEMNonTsEvent is a non-timestamped OEM-custom event.
+//
+// It holds 13 bytes of OEM-defined arbitrary data.
+type OEMNontsEvent struct {
```
Comment (rjoleary) [inline pkg/ipmi/structures.go:69]:
```suggestion
type OEMNonTsEvent struct {
```

File: pkg/ipmi/structures.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+type ChassisStatus struct {
+	CurrentPowerState byte
+	LastPowerEvent    byte
+	MiscChassisState  byte
+	FrontPanelButton  byte
+}
+
+type SELInfo struct {
```
Comment (rjoleary) [inline pkg/ipmi/structures.go:109]: Add comments for exported struct.

File: pkg/ipmi/ipmi_dev.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+type Dev struct {
+	*os.File
+}
+
+var (
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_dev.go:23]: Instead of these functions being globals, could they be private fields on the Dev struct?

**Comment by rminnich** [comment]
overall: be sure to golint in the pkg and other directory and go vet . -- they will complain about missing comments and that sort of thing. 

This is a big improvement, so let us know how we can help further to get it done.

File: pkg/ipmi/ipmi_dev.go (hunk 14 above)
Comment (ChriMarMe) [reply pkg/ipmi/ipmi_dev.go:23]: Your idea as well as my implementation have their own drawbacks. Yours is more logical on a structural level, but prolly not so nice to read and mine is the opposite. not so logical on the structural level, but better to read.

I would assume both is not good go style, but maybe I miss something. I"m undecided. Maybe you have a good argument for your idea?

**Comment by ChriMarMe** [comment]
> overall: be sure to golint in the pkg and other directory and go vet . -- they will complain about missing comments and that sort of thing.
> 
> This is a big improvement, so let us know how we can help further to get it done.

I use that per default, but it explodes with u-root. ~1k annotations over the whole project. T_T

File: pkg/ipmi/ipmi_dev.go (hunk 14 above)
Comment (ChriMarMe) [reply pkg/ipmi/ipmi_dev.go:23]: Ok, never mind. I see the prob here with glob vars

**Comment by ChriMarMe** [comment]
Any other suggestions for improvements before I squash it all together?

**Comment by ChriMarMe** [comment]
This is still not ready. Needs reimplementation of tests in the desired pattern.

**Comment by hugelgupf** [comment]
This is also a big change. I'd feel a lot more comfortable if we

- had VM tests that make _some_ IPMI call using this library
- or if @rjoleary ran one of our internal IPMI-using tests (disk unlock) using this code patched in

**Comment by ChriMarMe** [comment]
The work for the first option is way too much, especially if you have such tests in place internally already.

File: pkg/ipmi/ipmi_test.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+	"testing"
+)
+
+func TestWatchdogRunning(t *testing.T) {
+	i := GetMockIPMI()
+	_, err := i.WatchdogRunning()
+	if err != nil {
+		t.Error(err)
+	}
+
```
Comment (rjoleary) [inline pkg/ipmi/ipmi_test.go:17]: Remove empty line

**Comment by rjoleary** [comment]
> This is also a big change. I'd feel a lot more comfortable if we
> 
>     * had VM tests that make _some_ IPMI call using this library
> 
>     * or if @rjoleary ran one of our internal IPMI-using tests (disk unlock) using this code patched in

It possible to test SEL with QEMU's BMC sim. See http://www.linux-kvm.org/images/7/76/03x08-Juniper-Corey_Minyard-UsingIPMIinQEMU.ods.pdf

File: pkg/ipmi/mock.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"os"
+	"syscall"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+type mock struct {
```
Comment (rjoleary) [inline pkg/ipmi/mock.go:15]: Put a comment here saying these are mocks for unit tests.

File: pkg/ipmi/mock.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+func (m *mock) GetFile() *os.File {
+	return nil
+}
+
+func GetMockIPMI() *IPMI {
+	return &IPMI{Handler: &mock{
+		unixSyscall:         nil,
+		fileSyscallConn:     nil,
+		fileSetReadDeadline: nil,
+		connRead:            nil,
```
Comment (rjoleary) [inline pkg/ipmi/mock.go:45]: nil is the default

File: pkg/ipmi/linux.go [hunk 18]
```diff
@@ -1,7 +1,9 @@
-// Copyright 2019 the u-root Authors. All rights reserved
+// Copyright 2019-2021 the u-root Authors. All rights reserved
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+//go:build linux
```
Comment (rjoleary) [inline pkg/ipmi/linux.go:5]: Why rename this file from ipmi_linux.go?

**Comment by ChriMarMe** [comment]
This PR is tained somehow. I have to reopen i guess.

**Comment by ChriMarMe** [comment]
https://github.com/u-root/u-root/pull/2252

**Pull request closed without merging**